File: 42-trapping-rain-water/trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        l = 0
        n = len(height)
        r = n - 1
        leftmax = rightmax = 0
        ans = 0

        while l<=r:
            if height[l] < height[r]:
                if height[l] >= leftmax:
                    leftmax = height[l]
                else:
                    ans += leftmax - height[l]
                
                l += 1
            else:
                if height[r] >= rightmax:
                    rightmax = height[r]
                else:
                    ans += rightmax - height[r]
                r -= 1
                
            
        
        return ans
        

File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int lmax = height[0];
        int rmax = height[n - 1];
        int l = 0;
        int r = n - 1;
        int ans = 0;

        while (l < r){
            if (lmax <= rmax){
                l += 1;
                lmax = max(height[l], lmax);
                ans += lmax - height[l];
            }
            else{
                r -= 1;
                rmax = max(height[r], rmax);
                ans += rmax - height[r];
            }
        }

        return ans;
        
    }
};